python/tk_multi_publish2/api/manager.py

```diff
@@ -0,0 +1,444 @@
+            self._attach_plugins(new_file_items)
+
+            new_items.extend(new_file_items)
+
+        return new_items
+
+    def collect_session(self):
+        """
+        Run the collection logic to populate the graph with items to publish.
+
+        This will reestablish the state of the graph, recomputing everything.
+        Any externally added file paths will be retained.
+
+        :returns: A list of the items created.
+        """
+
+        # this will clear the graph of all non-persistent items.
+        self._graph.clear()
+
+        # get a list of all items in the graph prior to collection (this should
+        # be only the persistent items)
+        item_ids_before = [i.id for i in self._graph.items]
+
+        # we supply the root item of the graph for parenting of items that
+        # are collected.
+        self._collector_instance.run_process_current_session(
+            self._graph.root_item)
+
+        # get a list of all items in the graph after collection
+        item_ids_after = [i.id for i in self._graph.items]
+
+        # calculate which items are new
+        new_item_ids = list(set(item_ids_after) - set(item_ids_before))
+        new_items = self._graph.get_items_for_ids(new_item_ids)
+
+        # attach the appropriate plugins to the new items
+        if new_items:
+            self._attach_plugins(new_items)
+
+        return new_items
+
+    def load_graph(self, path):
+        """Deserialize and load a graph that was previously to disk.
+
+        NOTE: the manager's current graph will be lost.
+        """
+        self._graph = PublishGraph.load(path)
+
+    def save_graph(self, path):
+        """Serialize and save the underlying publish graph to disk."""
+        self._graph.save(path)
+
+    def validate(self):
+
+        all_valid = True
+
+        self.publish_logger.info("Running validation pass...")
+        for item in self._graph:
+            if not item.active:
+                self.publish_logger.debug(
+                    "Skipping item '%s' because it is inactive" % (item,))
+                continue
+
+            if not item.tasks:
+                self.publish_logger.debug(
+                    "Skipping item '%s' because it has no tasks attached." %
+                    (item,)
+                )
+                continue
+
+            self.publish_logger.debug("Validating: %s" % (item,))
+
+            for task in item.tasks:
+
+                if not task.active:
+                    self.publish_logger.debug(
+                        "Skipping inactive task: %s" % (task,))
+                    continue
+
+                self.publish_logger.debug(
+                    "Running validation for task: %s" % (task,))
+                if not task.validate():
+                    all_valid = False
+
+        return all_valid
+
+    def publish(self):
+
+        self.publish_logger.info("Executing publish pass...")
+
+        # TODO: re-validate?
```

Comment:
Not sure if validation should happen again here in the API. We do this on the UI side, but maybe for the UI we don't want to enforce a validate on every publish?

Reply:
I think it's fine to publish without doing validation in the API. Not everything needs to be validated all the time when using the API directly. Validation is there for a end-user. Devs are fine with an exception being raised I think.

Reply:
yes, validation should be a "TD-opt in" - we shouldn't force all workflows to be validated - but a TD should be able to enforce validation of input. 

If a user is writing a custom UI that replaces the publish UI, they can still run the validation programatically, right?